renderdoc/core/intervals.h

```diff
@@ -168,4 +167,31 @@ struct Intervals
     it--;
     return Wrap(it);
   }
-};
\ No newline at end of file
+};
+
+template <typename T, typename Compose>
+void updateIntervals(Intervals<T> &a, uint64_t begin, uint64_t end, T val, Compose comp)
+{
+  for(auto i = a.find(begin); i != a.end() && i.start() < end; i++)
```

Comment:
The comments for `find` above state that it finds the first interval that starts **after** the parameter. Is that definitely what we want here? In my mind this is basically taking an arbitrary range and then applying the compose function to all intervals that overlap it.

So if `begin` sits in the middle of an interval don't we want to conservatively over-estimate and apply the composition to the whole of that interval - even the part before `begin`, rather than under-estimate and apply to the next whole interval contained? Or do I misunderstand what the intention is here.

Reply:
`find(x)` "finds the interval containing `x`" (per the comment above the definition of `find`).

The comment you are referring to was meant to explain the implementation of `find` (find the interval after `x`, then step back one). Although now I'm trying to remember why I didn't just use `lower_bound` and skip the `it--`...

I will at least clean up that confusing comment, and see if the implementation of `find` can't also be simplified.

Reply:
Ah, my mistake. Yes that could do with a bit of clarification.

Also I know I've been very lax with this and only have recently added them, but the `Intervals` class as well as these helpers are good candidates for writing some unit tests. See for example `basic_types_tests.cpp`.

Reply:
I did have some unit tests when I first that class, but it looks like they were misplaced somewhere along the way. I will also add new tests for the helpers.